### PR TITLE
Remove dependency between dali_test.bin and dali_operators lib

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Description
- [ ] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [ ] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [x] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR

Removes dependecy between dali_test.bin and dali operators lib

#### Additional information
- Affected modules and functionalities:

C++ Tests
- Key points relevant for the review:

NA
### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [x] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/executor/executor_test.cc [hunk 1]
```diff
@@ -600,11 +590,16 @@ TYPED_TEST(ExecutorSyncTest, TestPrefetchedExecution) {
...
   ASSERT_TRUE(ws.OutputIsType<GPUBackend>(0));
   TensorList<GPUBackend> &res1 = ws.Output<GPUBackend>(0);
-  for (int i = 0; i < batch_size; ++i) {
-    this->VerifyDecode(
-        res1.template tensor<uint8>(i),
-        res1.tensor_shape(i)[0],
-        res1.tensor_shape(i)[1], i);
+
+  TensorList<CPUBackend> res_cpu;
+  res_cpu.Copy(res1, cudaStream_t(0));
```
Comment (JanuszL) [inline dali/pipeline/executor/executor_test.cc:595]:
```suggestion
  res_cpu.Copy(res1);
```
I think you can use just the default ordering.

File: dali/pipeline/executor/executor_test.cc [hunk 2]
```diff
@@ -616,12 +611,13 @@ TYPED_TEST(ExecutorSyncTest, TestPrefetchedExecution) {
...
   ASSERT_EQ(cb_counter, 2);
   TensorList<GPUBackend> &res2 = ws.Output<GPUBackend>(0);
-  for (int i = 0; i < batch_size; ++i) {
-    this->VerifyDecode(
-        res2.template tensor<uint8>(i),
-        res2.tensor_shape(i)[0],
-        res2.tensor_shape(i)[1],
-        i+batch_size);
+  TensorList<CPUBackend> res_cpu2;
+  res_cpu2.Copy(res2, cudaStream_t(0));
```
Comment (JanuszL) [inline dali/pipeline/executor/executor_test.cc:615]: Same:
```suggestion
  res_cpu2.Copy(res2);
```

File: dali/pipeline/pipeline_test.cc [hunk 3]
```diff
@@ -533,51 +524,30 @@ TYPED_TEST(PipelineTest, TestSeedSet) {
...
+  void CheckResults(Pipeline &pipe, int batch_size, int Iter, TensorList<CPUBackend> &data) {
     DeviceWorkspace ws;
     pipe.Outputs(&ws);
     ASSERT_EQ(ws.NumOutput(), 1);
     ASSERT_EQ(ws.NumInput(), 0);
     ASSERT_TRUE(ws.OutputIsType<GPUBackend>(0));
     TensorList<GPUBackend> &res1 = ws.Output<GPUBackend>(0);
+
+    TensorList<CPUBackend> res_cpu;
+    res_cpu.Copy(res1, cudaStream_t(0));
```
Comment (JanuszL) [inline dali/pipeline/pipeline_test.cc:539]:
```suggestion
    res_cpu.Copy(res1);
```

File: dali/pipeline/pipeline_test.cc [hunk 4]
```diff
@@ -533,51 +524,30 @@ TYPED_TEST(PipelineTest, TestSeedSet) {
...
+    auto data_view = view<uint8_t>(data);
     for (int j = 0; j < batch_size; ++j) {
-      this->VerifyDecode(
-          res1.template tensor<uint8>(j),
-          res1.tensor_shape(j)[0],
-          res1.tensor_shape(j)[1], (Iter * batch_size + j));
+      int data_idx = (Iter * batch_size + j) % data_view.shape.num_samples();
+      Check(res_cpu_view[j], data_view[data_idx]);
     }
   }
```
Comment (JanuszL) [inline dali/pipeline/pipeline_test.cc:548]: Maybe this can be extracted to a utility function (copy to the cpu and compare with the reference)?

Comment (JanuszL) [reply dali/pipeline/pipeline_test.cc:548]: Ok, it is already is. Can you use it inside `executor_test.cc`?

File: dali/test/dali_test_utils.cc [hunk 5]
```diff
@@ -12,21 +12,85 @@
...
+  auto res_cpu_view = view<uint8_t>(res_cpu);
+  auto data_view = view<uint8_t>(data);
+  for (int j = 0; j < batch_size; ++j) {
+    int data_idx = (i * batch_size + j) % data_view.shape.num_samples();
+    Check(res_cpu_view[j], data_view[data_idx]);
+  }
+}
 
 std::string CurrentExecutableDir() {
-    constexpr std::size_t PATH_MAX = 4096;
```
Comment (jantonguirao) [inline dali/test/dali_test_utils.cc:23]: PATH_MAX collides with a linux sys header

File: dali/test/dali_test_utils.h [hunk 6]
```diff
@@ -16,16 +16,42 @@
...
   return std::to_string(static_cast<std::underlying_type_t<Enum>>(value));
 }
 
 namespace dali {
+namespace test {
 
 std::string CurrentExecutableDir();
 
+/**
+ * @brief Produces a batch of ND random data with variable shapes
```
Comment (JanuszL) [inline dali/test/dali_test_utils.h:35]: Please add param description as well.

**Review by JanuszL — APPROVED** [review]

**Review by JanuszL — APPROVED** [review]

File: dali/test/dali_plugin_manager_test.cc [hunk 7]
```diff
@@ -16,28 +16,45 @@
...
+  static const std::string plugin_lib =
+      dali::test::CurrentExecutableDir() + "/libcustomdummyplugin.so";
+  return plugin_lib;
 }
 
 namespace other_ns {
 
-template <typename ImgType>
-class DummyTest : public ::dali::GenericMatchingTest<ImgType> {
+class DummyTest : public ::dali::DALITest {
```
Comment (klecki) [inline dali/test/dali_plugin_manager_test.cc:29]: Nitpick, the GenericMatchingTest tested both CPU and GPU variants of the op. 
Now we just invoke the CPU one. I just want to check if this is fine?

File: dali/pipeline/executor/executor_test.cc [hunk 8]
```diff
@@ -593,36 +584,12 @@ TYPED_TEST(ExecutorSyncTest, TestPrefetchedExecution) {
...
+  test::CheckResults(ws, batch_size, 0, tl);
 
   exe->Outputs(&ws);
-  ASSERT_EQ(ws.NumOutput(), 1);
-  ASSERT_EQ(ws.NumInput(), 0);
-  ASSERT_TRUE(ws.OutputIsType<GPUBackend>(0));
-
-  auto status_2 = barrier_future_2.wait_for(std::chrono::seconds(5));
-  ASSERT_EQ(status_2, std::future_status::ready);
-  ASSERT_EQ(cb_counter, 2);
```
Comment (klecki) [inline dali/pipeline/executor/executor_test.cc:617]: Should we really get rid of those cb_counter checks? (I don't think it sees much use, but it is still some kind of feature).

**Review by klecki — COMMENTED** [review]
Found a bug with ndim, and I think too much deleted code with the callback test. The rest are mostly suggestions.

File: dali/test/dali_test_utils.cc [hunk 9]
```diff
@@ -12,15 +12,25 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <unistd.h>
+#include "dali/test/dali_test_utils.h"
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/dali_test_utils.cc:15]: the first include can (and it's advised to be) the header associated with this cc file. That is my understanding

File: dali/test/dali_test_utils.cc [hunk 10]
```diff
@@ -29,4 +39,44 @@ std::string CurrentExecutableDir() {
     return {};
 }
 
+void MakeRandomBatch(TensorList<CPUBackend> &data, int N, int ndim,
+                     int64_t min_extent, int64_t max_extent) {
+  std::mt19937_64 rng(1234);
+  TensorListShape<3> tl_sh(N);
```
[the remark this answers is not in the document]
Comment (jantonguirao) [reply dali/test/dali_test_utils.cc:45]: will fix

File: dali/test/dali_plugin_manager_test.cc [hunk 11]
```diff
@@ -27,34 +27,47 @@ static const std::string& DummyPluginLibPath() {
...
+  void TestPlugin(const std::string &backend) {
+    LoadDummyPlugin();
+
+    dali::TensorList<dali::CPUBackend> data;
+    dali::test::MakeRandomBatch(data, 3);
 
-TEST_F(DummyTest, PluginShouldBeUsable) {
-  LoadDummyPlugin();
+    std::array<std::string, 2> backends = {"cpu", "gpu"};
+    for (auto backend : backends) {
```
Comment (klecki) [inline dali/test/dali_plugin_manager_test.cc:42]: This for loop does the same thing as the `this->TestPlugin("cpu"/"gpu");`.
You can remove it.

**Review by klecki — APPROVED** [review]
Please remove the for loop, otherwise ok.

**Pull request merged**